src/serialize.rs

```diff
@@ -137,19 +150,36 @@ impl EGraph {
         )> = Vec::new();
         let mut functions_kept = 0usize;
         let mut let_bindings = HashMap::default();
+        // Track all e-classes that appear as outputs of any function (even truncated/discarded ones)
+        let mut eclasses_with_outputs = HashSet::default();
         for (name, function) in self.functions.iter() {
             if functions_kept >= max_functions {
                 discarded_functions.push(name.clone());
+                // Track outputs of discarded functions
+                self.backend.for_each_while(function.backend_id, |row| {
```

Comment:
I think it's ok to not count as empty e-classes that were omitted due to max limits as well. 

Basically like once we reach max functions or nodes, we can treat any pointers to those e-classes as omitted not empty. 

The key thing is making sure that if we aren't having max nodes then empty e-classes are properly omitted.